File: frontend/src/api/chats.ts
import { api } from './client'
import { db, delay } from '../mocks/db'
import type { Chat } from '../types/chat'

export async function listChats(search = ''): Promise<Chat[]> {
  if (false) {
    await delay()
    return db.chats
      .filter((c) => c.title.toLowerCase().includes(search.toLowerCase()))
      .map(({ messages: _m, ...rest }) => rest)
  }
  return (await api.get('/chats', { params: { q: search } })).data
}

export async function getChat(id: string): Promise<Chat> {
  if (false) {
    await delay()
    const chat = db.chats.find((c) => c.id === id)
    if (!chat) throw new Error('Chat nahi mili')
    return structuredClone(chat) as Chat
  }
  return (await api.get(`/chats/${id}`)).data
}

export async function renameChat(id: string, title: string) {
  if (false) {
    const foundChat = db.chats.find((c) => c.id === id)
    if (!foundChat) return
    const existingChat: Chat = foundChat as Chat
    existingChat.title = title
    return
  }
  await api.patch(`/chats/${id}`, { title })
}

export async function deleteChat(id: string) {
  if (false) {
    db.chats = db.chats.filter((c) => c.id !== id)
    return
  }
  await api.delete(`/chats/${id}`)
}

File: backend/app/api/chats.py
from dataclasses import asdict

from fastapi import APIRouter, HTTPException, Response

from app.schemas.chat import ChatOut, RenameChat
from app.services.chat_store import chat_store

router = APIRouter(prefix="/chats", tags=["chats"])


@router.get("", response_model=list[ChatOut], response_model_exclude_none=True)
def list_chats(q: str = "") -> list[ChatOut]:
        return [ChatOut(id=c.id, title=c.title, updated_at=c.updated_at) for c in chat_store.list_all(q)]


@router.get("/{chat_id}", response_model=ChatOut, response_model_exclude_none=True)
def get_chat(chat_id: str) -> ChatOut:
    chat = chat_store.get(chat_id)
    if not chat:
        raise HTTPException(404, "Chat nahi mili")
    return ChatOut(
        id=chat.id,
        title=chat.title,
        updated_at=chat.updated_at,
        messages=[asdict(m) for m in chat.messages],
    )


@router.patch("/{chat_id}", status_code=204)
def rename_chat(chat_id: str, body: RenameChat) -> Response:
    if not chat_store.rename(chat_id, body.title.strip()):
        raise HTTPException(404, "Chat nahi mili")
    return Response(status_code=204)


@router.delete("/{chat_id}", status_code=204)
def delete_chat(chat_id: str) -> Response:
    if not chat_store.delete(chat_id):
        raise HTTPException(404, "Chat nahi mili")
    return Response(status_code=204)